[DPE-4583] - docs: add CC tutorial (#123)
canonical/kafka-k8s-operator: docs/tutorial/t-reassign-partitions.md

As the recommendation for production is to deploy a separate Juju application for running CruiseControl, we will cover that

```shell
juju deploy kafka-k8s --config roles=balancer -n 1 cruise-control
```

Earlier in the tutorial, we covered enabling TLS encryption, so we will repeat that step here for the new `cruise-control` application

```shell
juju relate cruise-control:certificates self-signed-certificates
```

Now, to make the new `cruise-control` application aware of the existing Kafka cluster, we will relate the two applications using the `peer_cluster` relation interface, ensuring that the `broker` cluster is using the `peer-cluster` relation-endpoint, and the `balancer` cluster is using the `peer-cluster-orchestrator` relation-endpoint

```shell
juju relate kafka-k8s:peer-cluster cruise-control:peer-cluster-orchestrator
```

### Adding new brokers

> Let's assume that we have three `kafka-k8s` units already, and have an active client application writing messages to an existing topic (TODO: link to t-relate-kafka.md). Let's scale the `kafka-k8s` application out to four units

Author: 🤦🏾 
Thank you!

  "logDirs": [
    {
      "partitions": [
        {
          "partition": "__KafkaCruiseControlModelTrainingSamples-10",
          "size": 0,
          "offsetLag": 0,
          "isFuture": false
        },
        ...
```

### Removing old brokers

Following on from the previous example of `Adding new brokers`, let's proceed to safely remove the new broker unit. Before scaling down the Juju application, we must make sure to carefully move any existing data on units about to be removed, to another unit, so as to preserve data availability.

In practice, this means running a `rebalance` Juju action as seen above, **BEFORE** scaling down the appilcation. This ensures that data is moved, prior to the unit becoming unreachable and permanently losing the data on it.

> **NOTE** - As partition data is replicated across a finite number of units based on the Kafka cluster's configured `replication.factor` property (usually == 3), it is imperative to remove only 1 broker at a time, so as to avoid losing all available replicas for a given partition

To remove the most recent broker unit 3 from the previous example, re-run the action with a new parameter value of `--mode=remove`

```shell

> juju run cruise-control/0 rebalance mode=remove dryrun=false brokerid=3 --wait=10m

Reviewer: Is it `--mode=remove` as above or `mode=remove` as here?

```

> This does not remove the unit, but moves the partitions from the current broker unit to other brokers within the cluster.

Reviewer: ```suggestion
This does not remove the unit, but moves the partitions from the broker unit number `3`  to other brokers within the cluster.
```

> Once the action has completed, we can verify that broker 3 no longer has any assigned partitions by doing the following

Reviewer: ```suggestion
Once the action has been completed, we can verify that broker `3` no longer has any assigned partitions by doing the following:
```

```bash
juju ssh --container kafka kafka-k8s/leader \
    '/opt/kafka/bin/kafka-log-dirs.sh' \
    '--describe' \
    '--bootstrap-server localhost:9093' \
    '--command-config /etc/kafka/client.properties' \
    | tail -n +3 | jq -c '.brokers[] | select(.broker == 3)' | jq
```

> This should result something similar to the result seen below, with broker 3 now having no partitions assigned, completing the safe reassignment of a broker's data to other brokers in the cluster

Reviewer: ```suggestion
This should produce output similar to the result seen below, with broker `3` now having no partitions assigned, completing the safe reassignment of a broker's data to other brokers in the cluster:
```

```json
{
  "broker": 3,
  "logDirs": [
    {
      "partitions": [],
      "error": null,
      "logDir": "/var/lib/kafka/data"
    }
  ]
}
```

> Now, it is safe to scale-in the cluster, removing broker 3

Reviewer: ```suggestion
Now, it is safe to scale-in the cluster, removing broker number `3`:
```

```shell

> juju scale-application kafka-k8s 3

Reviewer 2: Ok, on all my tests, the largest unitid is removed when we scale down. I think this behaviour is worth mentioning

> juju scale-application kafka-k8s 3

Author: I want to keep the docs and UX as similar for VM and K8s as possible.
You're right, for K8s StatefulSets, you can only scale-in from the highest ordinal unit, but I don't think that should change what we communicate here.

Do you think it's important enough to add a change to the K8s feature here, something like:
```python
if substrate == "k8s" and event.params['brokerid'] != max([broker.id for broker in self.charm.state.brokers]) and event.params['mode'] == "remove":
    event.fail("trying to remove non-highest unit")
```
?

> juju scale-application kafka-k8s 3

Reviewer 2: I think we can open a bug ticket and call it a day as far as this PR is concerned

```

### Full cluster rebalancing

> If a cluster has been running in production for some time, partition allocations may be imbalanced. This can happen naturally as topic load varies, topics are reconfigured to have more/less partitions, or new topics are added/removed etc. As such, it is expected that periodically during the lifetime of a deployment, an administrator will be called upon to better distribute the partition allocation across the existing broker units as part of cluster maintenance.

Reviewer: ```suggestion
If a cluster has been running in production for some time, partition allocations may be imbalanced. This can happen naturally as topic load varies, topics are reconfigured to have more/less partitions, or new topics are added/removed etc. As such, it is expected that periodically during the lifetime of a deployment, an administrator will run a command to redistribute the partition allocation across the existing broker units as part of cluster maintenance.
```

> Different to `Adding new brokers` or `Removing old brokers`, this includes a full re-shuffle of partition allocation across all currently live broker units.

Reviewer: ```suggestion
Unlike `Adding new brokers` or `Removing old brokers`, this includes a full re-shuffle of partition allocation across all currently live broker units.
```

> To achieve this, re-run the action with a new parameter value of `mode=full`, as a `dryrun` for now

Reviewer: ```suggestion
To achieve this, re-run the action with a new parameter value of `mode=full`. Let's try it with a `dryrun` (by default) for now:
```

```shell
juju run cruise-control/0 rebalance mode=full --wait=10m
```

> Looking to the bottom of the action result, we can see a measure of the `balancedness` score, and how the new proposal improves on the current balance

Reviewer: ```suggestion
Looking at the bottom of the output, we can see the value of the `balancedness` score before and after full rebalance:
```

```
summary:
  ...
  ondemandbalancednessscoreafter: "90.06926434109423"
  ondemandbalancednessscorebefore: "85.15942156660185"
  ...
```

> As before, to execute the proposal, run

Reviewer: ```suggestion
To implement the proposed changes, run:
```
